File: src/leetcode/python/medium/PeekingIterator_284.py
class Iterator(object):
    def __init__(self, nums):
        """
        Initializes an iterator object to the beginning of a list.
        :type nums: List[int]
        """

    def hasNext(self):
        """
        Returns true if the iteration has more elements.
        :rtype: bool
        """

    def next(self):
        """
        Returns the next element in the iteration.
        :rtype: int
        """


class PeekingIterator_284(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.iter = iterator
        self.hasElement = False
        self.integer = None

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        if not self.hasElement:
            self.integer = self.iter.next()
            self.hasElement = True
        return self.integer

    def next(self):
        """
        :rtype: int
        """
        if self.hasElement:
            self.hasElement = False
            return self.integer
        return self.iter.next()

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.hasElement or self.iter.hasNext()

File: src/leetcode/java/PeekingIterator_284.java
package leetcode.java;

import java.util.Iterator;

public class PeekingIterator_284 implements Iterator<Integer> {
  private Iterator<Integer> iter;
  private boolean hasElement;
  private Integer element;

  public PeekingIterator_284(Iterator<Integer> iterator) {
    // initialize any member here.
    this.iter = iterator;
    this.hasElement = false;
  }

  // Returns the next element in the iteration without advancing the iterator.
  public Integer peek() {
    if (!hasElement) {
      element = iter.next();
      hasElement = true;
    }
    return element;
  }

  // hasNext() and next() should behave the same as in the Iterator interface.
  // Override them if needed.
  @Override
  public Integer next() {
    if (hasElement) {
      hasElement = false;
      return element;
    }
    return iter.next();
  }

  @Override
  public boolean hasNext() {
    return hasElement || iter.hasNext();
  }
}
